test/unit/context/test-node-values.js

```diff
@@ -0,0 +1,878 @@
+      expect(sibling1Stub).to.be.calledTwice.calledWith('(A|Z)C');
+      expect(sibling2Stub).to.be.calledTwice.calledWith('(A|Z)');
+      expect(cousin1Stub).to.be.calledTwice.calledWith('(A|Z)');
+      expect(calcSpy).to.have.callCount(5);
+    });
+  });
+
+  describe('connected, with deps', () => {
+    let sibling1, sibling1Stub;
+    let sibling2, sibling2Stub;
+    let cousin1, cousin1Stub;
+    let parent, parentStub;
+    let grandparent, grandparentStub;
+
+    beforeEach(async () => {
+      doc.body.appendChild(tree);
+      sibling1 = ContextNode.get(el('T-1-1-1'));
+      sibling2 = ContextNode.get(el('T-1-1-2'));
+      cousin1 = ContextNode.get(el('T-1-2-1'));
+      parent = ContextNode.get(el('T-1-1'));
+      grandparent = ContextNode.get(el('T-1'));
+
+      sibling1Stub = sandbox.stub();
+      sibling2Stub = sandbox.stub();
+      cousin1Stub = sandbox.stub();
+      parentStub = sandbox.stub();
+      grandparentStub = sandbox.stub();
+
+      await waitForDiscover(grandparent, parent, sibling1, sibling2, cousin1);
+
+      sibling1.values.subscribe(Computed, sibling1Stub);
+      sibling2.values.subscribe(Computed, sibling2Stub);
+      cousin1.values.subscribe(Computed, cousin1Stub);
+      parent.values.subscribe(Computed, parentStub);
+      grandparent.values.subscribe(Computed, grandparentStub);
+      clock.runAll();
+      calcSpy.resetHistory();
+    });
+
+    it('should compute with all dependencies only', () => {
+      expect(grandparentStub).to.not.be.called;
+      expect(parentStub).to.not.be.called;
+      expect(sibling1Stub).to.not.be.called;
+      expect(sibling2Stub).to.not.be.called;
+      expect(cousin1Stub).to.not.be.called;
+
+      // Add a computed recursive dependency.
+      grandparent.values.set(Concat, 'OWNER1', 'A');
+      clock.runAll();
+      expect(grandparentStub).to.not.be.called;
+      expect(parentStub).to.not.be.called;
+      expect(sibling1Stub).to.not.be.called;
+      expect(sibling2Stub).to.not.be.called;
+      expect(cousin1Stub).to.not.be.called;
+
+      // Add a recursive dependency.
+      grandparent.values.set(Recursive, 'OWNER1', 'recursive');
+      clock.runAll();
+      expect(grandparentStub).to.not.be.called;
+      expect(parentStub).to.not.be.called;
+      expect(sibling1Stub).to.not.be.called;
+      expect(sibling2Stub).to.not.be.called;
+      expect(cousin1Stub).to.not.be.called;
+
+      // Add a non-recursive dependency.
+      calcSpy.resetHistory();
+      parent.values.set(NonRecursive, 'OWNER1', 'non-recursive');
+      clock.runAll();
+      expect(parentStub).to.be.calledOnce.calledWith(
+        'no-input/non-recursive/recursive/A'
+      );
+      expect(grandparentStub).to.not.be.called;
+      expect(sibling1Stub).to.not.be.called;
+      expect(sibling2Stub).to.not.be.called;
+      expect(cousin1Stub).to.not.be.called;
+      // One computation for `NonRecursive` and one for `Computed`.
+      expect(calcSpy).to.have.callCount(2);
```

Comment:
Can we have a check where appropriate? I'm not sure what triggers a calc and what doesn't.

Reply:
Can you clarify? The `calcSpy` is mostly as a control against too-many recalculations. The actual values and handlers are always checked for each scenario. It's just that in some cases one change could rippled through and do many recomputes that are unnecessary. So I always include this check to control for this.

Reply:
I see. But no - there's no way to easily separate them in this model. There could be many reasons to recompute, but they are all deduplicated and fewer computes are run. But they are all pushed through the same queue, so the exact sequence is not visible to callers.